llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -547,15 +555,36 @@ static string_vector saveDeviceImageProperty(
 
     {
       Module *M = ResultModules[I].get();
+      bool HasIndirectlyCalledAssert = false;
+      std::vector<llvm::Function *> Kernels;
       for (auto &F : M->functions()) {
         // TODO: handle SYCL_EXTERNAL functions for dynamic linkage.
         // TODO: handle function pointers.
         if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
```

Comment:
What about inserting an early exit here?

Additionally, since we have cache mechanism, we might want to call `hasAssertInFunctionCallGraph` for every function, but only use it result if function is a kernel - that way we might populate the cache earlier and we can also find `referenced-indirectly` function earlier which will save us some time (because we would be able to skip the analysis of the rest of functions)

Reply:
Applied an early exit here.

About the caching mechanism: now it's tied to the current graph traversal. This optimization will work as you expect only if we construct full call graph as a structure (with nodes and details).
I believe we can do this and apply this optimization in future, or with another way, but it'll require a separate PR.